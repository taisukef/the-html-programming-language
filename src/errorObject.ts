export class SyntaxError extends Error {
  readonly node: Node;
  constructor(message: string, node: Node) {
    super(message);
    this.node = node;
  }
}

export class RuntimeError extends Error {
  readonly node: Node;
  constructor(message: string, node: Node) {
    super(message);
    this.node = node;
  }
}
