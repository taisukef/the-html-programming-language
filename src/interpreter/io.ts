export type IO = {
  output: (text: string) => void;
};
