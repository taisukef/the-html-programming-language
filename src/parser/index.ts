import { Program } from "../ast/index";
import { Statement } from "../ast/statement";
import { expectNothing } from "./expect";
import { parseStatement } from "./parseStatement";

export function parseProgram(program: HTMLElement): Program {
  const statements: Statement[] = [];

  let nodes: Node[] = Array.from(program.childNodes);
  while (nodes.length > 0) {
    const res = parseStatement(nodes);
    if (!res) {
      break;
    }
    const [stmt, next] = res;
    statements.push(stmt);
    nodes = next;
  }
  expectNothing(nodes);

  return {
    statements,
  };
}
