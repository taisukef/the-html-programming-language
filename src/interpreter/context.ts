import { IO } from "./io";

export type InterpreterContext = {
  io: IO;
};
