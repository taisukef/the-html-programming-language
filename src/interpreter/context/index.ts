import { IO } from "../io";
import { Environment } from "./environment";
import { Input } from "./input";

export type InterpreterContext = {
  io: IO;
  input: Input;
  environment: Environment;
};
