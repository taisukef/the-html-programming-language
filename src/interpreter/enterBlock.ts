import { Statement } from "../ast/statement";
import { createScope, Environment } from "./context/environment";
import { createFunctionValue } from "./value";

/**
 * Enter a block.
 * @returns a new environment
 */
export function enterBlock(
  block: readonly Statement[],
  environment: Environment
): Environment {
  const newScope = createScope();
  const newEnvironment: Environment = {
    scopes: [...environment.scopes, newScope],
  };
  // hoist declarations
  for (const statement of block) {
    switch (statement.type) {
      case "SectionDeclaration": {
        const func = createFunctionValue(newEnvironment, statement.body);
        newScope.bindings.set(statement.title, func);
        break;
      }
    }
  }
  return newEnvironment;
}
