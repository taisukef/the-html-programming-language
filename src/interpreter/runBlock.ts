import { Statement } from "../ast/statement";
import { createScope, Environment } from "./context/environment";
import { InterpreterContext } from "./context/index";
import { runStatement } from "./runStatement";
import { createFunctionValue, Value } from "./value";

export type RunBlockResult =
  | {
      readonly type: "normal";
    }
  | {
      readonly type: "footer";
      readonly value: Value;
    };

export async function runBlock(
  block: readonly Statement[],
  context: InterpreterContext
): Promise<RunBlockResult> {
  const newEnvironment = enterBlock(block, context.environment);
  const newContext = {
    ...context,
    environment: newEnvironment,
  };
  for (const statement of block) {
    const res = await runStatement(statement, newContext);
    if (res.type === "footer") {
      return {
        type: "footer",
        value: res.value,
      };
    }
  }
  return {
    type: "normal",
  };
}

/**
 * Enter a block.
 * @returns a new environment
 */
function enterBlock(
  block: readonly Statement[],
  environment: Environment
): Environment {
  const newScope = createScope();
  const newEnvironment: Environment = {
    scopes: [...environment.scopes, newScope],
  };
  // hoist declarations
  for (const statement of block) {
    switch (statement.type) {
      case "SectionDeclaration": {
        const func = createFunctionValue(newEnvironment, statement.body);
        newScope.bindings.set(statement.title, func);
        break;
      }
    }
  }
  return newEnvironment;
}
