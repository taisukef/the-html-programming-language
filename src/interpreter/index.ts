import { Program } from "../ast/index";
import { createEnvironment } from "./context/environment";
import { InterpreterContext } from "./context/index";
import { IO } from "./io";
import { runBlock } from "./runBlock";

export type Interpreter = {
  run: (program: Program) => Promise<void>;
};

export function createInterpreter(io: IO): Interpreter {
  return {
    async run(program) {
      const context: InterpreterContext = {
        io,
        environment: createEnvironment(),
      };
      await runBlock(program.statements, context);
    },
  };
}
