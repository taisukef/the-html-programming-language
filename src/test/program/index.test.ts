import { readdirSync, readFileSync } from "fs";
import path from "path";
import { RuntimeError, SyntaxError } from "../../errorObject";
import { createInterpreter } from "../../interpreter/index";
import { parseProgram } from "../../parser/index";

describe("program", () => {
  // test to run
  const caseFiles = readdirSync(path.join(__dirname, "cases"));

  for (const file of caseFiles) {
    const filePath = path.resolve(__dirname, "cases", file);
    if (!filePath.endsWith(".html")) {
      continue;
    }
    const testcaseName = path.basename(filePath);
    let content = readFileSync(filePath, "utf8").trim();
    // treat initial `//` lines as input
    let input = "";
    let match;
    while ((match = content.match(/^\/\/(.*)\n/))) {
      input += match[1] + "\n";
      content = content.slice(match[0].length);
    }

    it(testcaseName, async () => {
      document.body.innerHTML = content;
      let output = "";
      let resultError = null;
      try {
        const program = parseProgram(document.body);
        const interpreter = createInterpreter({
          output: (text) => {
            output += text;
          },
        });
        await interpreter.run(program);
      } catch (err) {
        resultError = err;
      }
      if (resultError) {
        expect(errorSnapshot(resultError)).toMatchSnapshot();
      } else {
        expect(output).toMatchSnapshot();
      }
    });
  }
});

function errorSnapshot(error: unknown) {
  if (error instanceof SyntaxError) {
    return {
      error: "SyntaxError",
      message: error.message,
      node: error.node,
    };
  } else if (error instanceof RuntimeError) {
    return {
      error: "RuntimeError",
      message: error.message,
      node: error.node,
    };
  }
  return error;
}
