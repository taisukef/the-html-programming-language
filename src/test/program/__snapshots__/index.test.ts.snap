// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`program dl.html 1`] = `"Hello, world!"`;

exports[`program fizzbuzz.html 1`] = `
"1
2
Fizz
4
Buzz
Fizz
7
8
Fizz
Buzz
11
Fizz
13
14
FizzBuzz
16
17
Fizz
19
Buzz
Fizz
22
23
Fizz
Buzz
26
Fizz
28
29
FizzBuzz
31
32
Fizz
34
Buzz
Fizz
37
38
Fizz
Buzz
41
Fizz
43
44
FizzBuzz
46
47
Fizz
49
Buzz
Fizz
52
53
Fizz
Buzz
56
Fizz
58
59
FizzBuzz
61
62
Fizz
64
Buzz
Fizz
67
68
Fizz
Buzz
71
Fizz
73
74
FizzBuzz
76
77
Fizz
79
Buzz
Fizz
82
83
Fizz
Buzz
86
Fizz
88
89
FizzBuzz
91
92
Fizz
94
Buzz
Fizz
97
98
Fizz
Buzz
"
`;

exports[`program func-no-arg-no-return-value.html 1`] = `"hello, world!"`;

exports[`program func-no-arg-return-value.html 1`] = `"hello, world!"`;

exports[`program func-one-arg-return-value.html 1`] = `"Hello, world!"`;

exports[`program func-two-args.html 1`] = `"hello, world!"`;

exports[`program hello_br_world.html 1`] = `
"Hello,
world!"
`;

exports[`program helloworld.html 1`] = `"Hello, world!"`;

exports[`program input-from-var.html 1`] = `"1368"`;

exports[`program input-reverse.html 1`] = `"hello, world!"`;

exports[`program math-many.html 1`] = `
"minus
8
times
30
divide
9
rem
1
gcd
6
min
-5
max
10
"
`;

exports[`program math-many2.html 1`] = `
"eq
1
0
neq
1
0
lt
1
0
gt
1
0
leq
1
0
geq
1
0
"
`;

exports[`program math-many3.html 1`] = `
"and
1
0
or
1
0
xor
1
0
implies
1
0
"
`;

exports[`program math-plus.html 1`] = `"6"`;

exports[`program ruby.html 1`] = `
"*
**
***
****
*****
"
`;

exports[`program var-and-a-func.html 1`] = `"Hello, world!"`;
